File: core-engine/src/main/java/com/ssetglow/spirelink/engine/LinkResolver.java
package com.ssetglow.spirelink.engine;

import com.ssetglow.spirelink.common.LinkEffectType;
import com.ssetglow.spirelink.common.SupportActionType;
import com.ssetglow.spirelink.domain.BattleCoordinationContext;
import com.ssetglow.spirelink.domain.LinkEffect;
import com.ssetglow.spirelink.domain.PlayerState;
import com.ssetglow.spirelink.domain.SupportActionRecord;

import java.util.ArrayList;
import java.util.List;

public class LinkResolver {
    public List<SupportActionRecord> triggerTurnStartLinks(BattleCoordinationContext context, String playerId, int currentTurn) {
        List<LinkEffect> toApply = new ArrayList<>();
        context.pendingLinks().removeIf(effect -> {
            boolean match = playerId.equals(effect.targetPlayerId()) && effect.expireTurn() >= currentTurn;
            if (match) {
                toApply.add(effect);
            }
            return match;
        });

        List<SupportActionRecord> records = new ArrayList<>();
        for (LinkEffect effect : toApply) {
            PlayerState target = context.requirePlayer(playerId);
            switch (effect.effectType()) {
                case BLOCK -> {
                    target.gainBlock(effect.value());
                    records.add(SupportActionRecord.now(effect.sourcePlayerId(), playerId, effect.sourceCardId(), SupportActionType.GRANT_BLOCK, effect.value()));
                }
                case DRAW -> {
                    target.queueDraw(effect.value());
                    records.add(SupportActionRecord.now(effect.sourcePlayerId(), playerId, effect.sourceCardId(), SupportActionType.GRANT_DRAW, effect.value()));
                }
                case COST_REDUCTION -> {
                    target.reduceNextCardCost(effect.value());
                    records.add(SupportActionRecord.now(effect.sourcePlayerId(), playerId, effect.sourceCardId(), SupportActionType.REDUCE_COST, effect.value()));
                }
                case ATTACK_BONUS -> {
                    target.addAttackBonus(effect.value());
                    records.add(SupportActionRecord.now(effect.sourcePlayerId(), playerId, effect.sourceCardId(), SupportActionType.GRANT_ATTACK_BONUS, effect.value()));
                }
                case CLEANSE -> {
                    target.cleanseDebuff(effect.value());
                    records.add(SupportActionRecord.now(effect.sourcePlayerId(), playerId, effect.sourceCardId(), SupportActionType.CLEANSE_DEBUFF, effect.value()));
                }
                case TEAM_BLOCK, CUSTOM -> {
                    target.gainBlock(effect.value());
                    records.add(SupportActionRecord.now(effect.sourcePlayerId(), playerId, effect.sourceCardId(), SupportActionType.PROTECT, effect.value()));
                }
            }
            context.recordSupport(records.get(records.size() - 1));
            context.incrementLinkTriggeredThisTurn();
        }
        return records;
    }

    public LinkEffect createEffect(String cardId, String sourcePlayerId, String targetPlayerId,
                                   LinkEffectType type, int value, com.ssetglow.spirelink.common.TargetRule targetRule,
                                   int expireTurn, String description) {
        return new LinkEffect(cardId, sourcePlayerId, targetPlayerId, type, value, targetRule, expireTurn, description);
    }
}

File: src/Runtime/Systems/LinkResolver.cs
namespace SpireLink.Runtime.Systems;

public sealed class LinkResolver
{
    public ExecutionContext TriggerTurnStartLinks(string playerId)
    {
        var battle = SpireLinkRuntimeState.CurrentBattle;
        battle.IncrementTurn(playerId);
        battle.ResetTurnScopedCounters();

        var context = new ExecutionContext(battle, "TURN_START_LINKS", playerId, playerId, false);
        foreach (var payload in battle.DequeueLinksForTurnStart(playerId))
        {
            var record = ToSupportAction(payload, playerId);
            battle.RecordSupport(record);
            battle.IncrementLinkTriggeredThisTurn();
            context.AddLog($"TriggerLink {payload.PayloadType} from {payload.SourcePlayerId} -> {playerId} value={payload.Value}");
        }

        return context;
    }

    private static SupportActionRecord ToSupportAction(LinkPayload payload, string targetPlayerId)
    {
        return new SupportActionRecord(
            payload.SourcePlayerId,
            targetPlayerId,
            payload.CardId,
            ToSupportActionType(payload.PayloadType),
            payload.Value
        );
    }

    private static SupportActionType ToSupportActionType(LinkPayloadType payloadType)
    {
        return payloadType switch
        {
            LinkPayloadType.Block => SupportActionType.GrantBlock,
            LinkPayloadType.Draw => SupportActionType.GrantDraw,
            LinkPayloadType.CostReduction => SupportActionType.ReduceCost,
            LinkPayloadType.AttackBonus => SupportActionType.GrantAttackBonus,
            LinkPayloadType.Cleanse => SupportActionType.CleanseDebuff,
            LinkPayloadType.TeamBlock => SupportActionType.Protect,
            _ => SupportActionType.GrantBlock
        };
    }
}
